drivers/flash/flash_rpi_pico.c

```diff
@@ -0,0 +1,307 @@
+							size_t count) {
+	rom_connect_internal_flash_fn connect_internal_flash = (rom_connect_internal_flash_fn)
+					rom_func_lookup_inline(ROM_FUNC_CONNECT_INTERNAL_FLASH);
+	rom_flash_exit_xip_fn flash_exit_xip = (rom_flash_exit_xip_fn)
+						rom_func_lookup_inline(ROM_FUNC_FLASH_EXIT_XIP);
+	rom_flash_flush_cache_fn flash_flush_cache = (rom_flash_flush_cache_fn)
+						rom_func_lookup_inline(ROM_FUNC_FLASH_FLUSH_CACHE);
+
+	flash_init_boot2_copyout();
+
+	__compiler_memory_barrier();
+
+	connect_internal_flash();
+	flash_exit_xip();
+	flash_write_partial_internal(flash_offs, data, count);
+	flash_flush_cache();
+	flash_enable_xip_via_boot2();
+}
+
+static bool is_invalid_range(off_t offset, uint32_t size)
+{
+	return (offset + size) > FLASH_SIZE;
+}
+
+static int flash_rpi_read(const struct device *dev, off_t offset, void *data, size_t size)
+{
+	if (is_invalid_range(offset, size)) {
+		LOG_ERR("Read range exceeds the flash boundaries");
+		return -EINVAL;
+	}
+
+	if (size == 0) {
+		return 0;
+	}
+
+	memcpy(data, (uint8_t *)(CONFIG_FLASH_BASE_ADDRESS + offset), size);
+
+	return 0;
+}
+
+static int flash_rpi_write(const struct device *dev, off_t offset, const void *data, size_t size)
+{
+	uint32_t key;
+	size_t bytes_to_write;
+	uint8_t *_data = (uint8_t *)data;
+
+	if (is_invalid_range(offset, size)) {
+		LOG_ERR("Write range exceeds the flash boundaries. Offset=%#lx, Size=%u",
+				offset, size);
+		return -EINVAL;
+	}
+
+	if (size == 0) {
+		return 0;
+	}
+
+	key = irq_lock();
+
+	if (offset % PAGE_SIZE > 0) {
+		bytes_to_write = PAGE_SIZE - (offset % PAGE_SIZE);
+		flash_write_partial(offset, _data, bytes_to_write);
+		_data += bytes_to_write;
+		size -= bytes_to_write;
+		offset += bytes_to_write;
+	}
+
+	if (size >= PAGE_SIZE) {
+		bytes_to_write = size & ~(PAGE_SIZE - 1);
+		flash_range_program(offset, _data, bytes_to_write);
+		_data += bytes_to_write;
+		size -= bytes_to_write;
+		offset += bytes_to_write;
+	}
+
+	if (size > 0) {
+		flash_write_partial(offset, _data, size);
+	}
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+static int flash_rpi_erase(const struct device *dev, off_t offset, size_t size)
+{
+	uint32_t key;
+
+	if (is_invalid_range(offset, size)) {
+		LOG_ERR("Erase range exceeds the flash boundaries. Offset=%#lx, Size=%u",
+				offset, size);
+		return -EINVAL;
+	}
+
+	if ((offset % SECTOR_SIZE) || (size % SECTOR_SIZE)) {
+		LOG_ERR("Erase range is not a multiple of the sector size. Offset=%#lx, Size=%u",
+				offset, size);
+		return -EINVAL;
+	}
+
+	if (size == 0) {
+		return 0;
+	}
+
+	key = irq_lock();
+
+	flash_range_erase(offset, size);
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+static const struct flash_parameters *flash_rpi_get_parameters(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return &flash_rpi_parameters;
+}
+
+#if CONFIG_FLASH_PAGE_LAYOUT
+
+static const struct flash_pages_layout flash_rpi_pages_layout = {
+	.pages_count = FLASH_SIZE / PAGE_SIZE,
```

Comment:
Shouldn't this use `SECTOR_SIZE` instead of `PAGE_SIZE` ?

Reply:
> Here, "page" means the smallest erasable area on the flash device.
So you're right